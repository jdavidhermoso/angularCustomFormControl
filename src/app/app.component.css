.choicer-form-container {
  width: 80%;
  margin: 0 auto;
}
