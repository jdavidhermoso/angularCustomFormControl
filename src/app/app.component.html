<div class="choicer-form-container">
  <form [formGroup]="myForm">
    <app-choicer formControlName="choicer" [itemsList]="items" [icon]="selectedChoicerItemIcon" [selectionMode]="selectionMode"></app-choicer>
  </form>
</div>

<div class="selected-options-container">
  <p>
    Selected cities
  </p>
  <p>
    This element is just to illustrate that We are using the choicer-component as FormControl.
  </p>

  <ul>
      <p *ngFor="let item of choicer.value">
        {{item.title}}
        </p>
  </ul>
</div>


