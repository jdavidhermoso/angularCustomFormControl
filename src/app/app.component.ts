import { Component, OnInit } from '@angular/core';
import { FormControl, FormGroup } from '@angular/forms';
import { SelectionModes } from './components/choicer/selection-modes.enum';
import { Item } from './models/item.model';
import { items } from './mockedItems';

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.css']
})
export class AppComponent implements OnInit {
  myForm: FormGroup;
  choicer: FormControl;
  selectedChoicerItemIcon = 'check';
  items: Item[] = items;
  selectionMode: SelectionModes = SelectionModes.Multi;

  ngOnInit() {
    this.choicer = new FormControl({ value: [], disabled: false });
    this.myForm = new FormGroup({
      choicer: this.choicer
    });
  }
}


