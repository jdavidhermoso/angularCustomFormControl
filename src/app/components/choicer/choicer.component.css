.choicer-item {
  position: relative;
  max-width: 100%;
  cursor: pointer;
  user-select: none;
  margin: 10px;
}

.choicer-item-content {
  height: 100px;
  overflow: scroll;
}

@media (min-width: 768px) {
  .choicer-item-container {
    max-width: 25%;
    display: inline-block;
  } 
}

.choicer-item.disabled, .choicer-list.disabled .choicer-item {  
  color: rgba(0,0,0,.26);
  background-color: rgba(0,0,0,.12);
  cursor: not-allowed;
  box-shadow: none;
}

.choicer-selected-icon-container {    
  opacity: 0;
}

.choicer-selected-icon-container .mat-icon {
  font-size: 40px;
  height: 40px;
  width: 40px;
  line-height: 40px;
}

.choicer-item.selected .choicer-selected-icon-container {
  opacity: 1;
}

.choicer-item-image {
  min-height: 175px;
  height: 175px;
}
