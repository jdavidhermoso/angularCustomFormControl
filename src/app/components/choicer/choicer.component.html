<div class="choicer-list" [class.disabled]="disabled">
  <div class="choicer-item-container" *ngFor="let item of itemsList">
      <mat-card class="choicer-item"      
      (click)="onItemClicked(item)"
      [class.selected]="isItemSelected(item.id)" [class.disabled]="item.disabled">
<mat-card-header>

<div mat-card-avatar class="">
    <div class="choicer-selected-icon-container">
        <ng-container>
          <mat-icon color="accent">{{icon}}</mat-icon>
        </ng-container>
      </div>
</div>
<mat-card-title>{{ item.title }}</mat-card-title>
<mat-card-subtitle *ngIf="item.subtitle">{{item.subtitle}}</mat-card-subtitle>
</mat-card-header>
<img mat-card-image class="choicer-item-image" [src]="item.image" alt="Photo of a Shiba Inu">
<mat-card-content class="choicer-item-content">
{{item.description}}
</mat-card-content>
</mat-card>
  </div>
  
</div>

