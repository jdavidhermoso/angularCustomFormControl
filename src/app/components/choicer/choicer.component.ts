import { ChangeDetectorRef, Component, forwardRef, Input, OnInit } from '@angular/core';
import { ControlValueAccessor, NG_VALUE_ACCESSOR } from '@angular/forms';
import { Item } from '../../models/item.model';
import { SelectionModes } from './selection-modes.enum';

@Component({
  selector: 'app-choicer',
  templateUrl: './choicer.component.html',
  styleUrls: ['./choicer.component.css'],
  providers: [
    {
      provide: NG_VALUE_ACCESSOR,
      useExisting: forwardRef(() => ChoicerComponent),
      multi: true
    }]
})
export class ChoicerComponent implements ControlValueAccessor {
  @Input() selectedItems: Item[] = [];
  @Input() itemsList: Item[];
  @Input() disabled = false;
  @Input() icon = 'check';
  @Input() selectionMode: SelectionModes = SelectionModes.Multi;

  onChange = (items: Item[]) => {
  };

  onTouched = () => {
  };

  public get value(): Item[] {
    return this.selectedItems;
  }

  public onItemClicked(item: Item) {
    this.onTouched();
    if (!this.disabled && !item.disabled) {
      this.toggleItemState(item);
    }
  }

  private toggleItemState(item: Item) {
    const itemId = item.id;
    if (this.isItemSelected(itemId)) {
      this.removeItemFromSelection(itemId);
    } else if (this.canSelectItem()) {
      this.selectedItems.push(item);
    }

    this.onChange(this.value);
  }

  private canSelectItem() {
    return (this.selectionMode === SelectionModes.Single && this.selectedItems.length < 1) || this.selectionMode === SelectionModes.Multi;
  }

  private removeItemFromSelection(itemToRemoveId: number) {
    this.selectedItems = this.selectedItems.filter((selectedItem) => {
      return selectedItem.id !== itemToRemoveId;
    });
  }

  writeValue(items: Item[]) {
    this.selectedItems = items;
    this.onChange(this.value);
  }

  registerOnChange(fn: (items: Item[]) => void) {
    this.onChange = fn;
  }

  registerOnTouched(fn: () => void): void {
    this.onTouched = fn;
  }

  private isItemSelected(itemId: number) {
    return this.getItemByItemId(itemId);
  }

  private getItemByItemId(itemId: number) {
    return this.selectedItems.find((item: Item) => {
      return item.id === itemId;
    });
  }
}
