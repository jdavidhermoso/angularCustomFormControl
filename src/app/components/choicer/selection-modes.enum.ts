export enum SelectionModes {
  Single = 'single',
  Multi = 'multi'
}
