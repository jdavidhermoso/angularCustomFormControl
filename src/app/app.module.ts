import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { MatButtonModule, MatCardModule, MatIconModule, MatListModule } from '@angular/material';
import { BrowserModule } from '@angular/platform-browser';
import { NgModule } from '@angular/core';

import { AppComponent } from './app.component';
import { ChoicerComponent } from './components/choicer/choicer.component';

@NgModule({
  declarations: [
    AppComponent,
    ChoicerComponent
  ],
  imports: [
    BrowserModule,
    FormsModule,
    ReactiveFormsModule,
    MatButtonModule,
    MatCardModule,
    MatIconModule
  ],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule {
}
