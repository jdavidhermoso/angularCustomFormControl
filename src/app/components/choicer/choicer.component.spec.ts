import { CUSTOM_ELEMENTS_SCHEMA, forwardRef } from '@angular/core';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { FormsModule, NG_VALUE_ACCESSOR, ReactiveFormsModule } from '@angular/forms';
import { MatButtonModule, MatCardModule, MatIcon, MatIconModule } from '@angular/material';
import { AppComponent } from '../../app.component';
import { Item } from '../../models/item.model';

import { ChoicerComponent } from './choicer.component';

describe('ChoicerComponent', () => {
  let component: ChoicerComponent;
  let containerComponent: AppComponent;
  let fixture: ComponentFixture<ChoicerComponent>;
  let containerFixture: ComponentFixture<AppComponent>;
  const singleItem = {
    id: 1,
    title: 'Item 1 title',
    subtitle: 'Item 1 subtitle',
    description: '',
    image: '',
    disabled: false
  };
  const items: Item[] = [singleItem];

  const setComponentState = (isDisabled: boolean) => {
    component.disabled = isDisabled;
  };

  const setSelectedItem = (item: Item) => {
    component.selectedItems.push(item);
  };

  const setSingleItemState = (isDisabled: boolean) => {
    singleItem.disabled = isDisabled;
  };

  const setContainerComponentItems = (items: Item[]) => {
    containerComponent.items = items;
  };

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [AppComponent, ChoicerComponent],
      providers: [
        {
          provide: NG_VALUE_ACCESSOR,
          useExisting: forwardRef(() => component),
          multi: true
        }],
      imports: [
        FormsModule,
        ReactiveFormsModule,
        MatButtonModule,
        MatCardModule,
        MatIconModule]
    })
      .compileComponents();
  }));

  beforeEach(() => {
    containerFixture = TestBed.createComponent(AppComponent);
    containerComponent = containerFixture.componentInstance;
    fixture = TestBed.createComponent(ChoicerComponent);
    component = fixture.componentInstance;

    setContainerComponentItems(items);
    containerComponent.ngOnInit();

    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  describe('ChoicerComponent: Component or items are disabled.', () => {
    it('Component is disabled, so none element should be selected', () => {
      setComponentState(true);
      component.onItemClicked(singleItem);
      expect(component.selectedItems).toMatchSnapshot();
    });

    it('Single item is disabled, so none element should be selected', () => {
      setSingleItemState(true);
      setComponentState(false);

      component.onItemClicked(singleItem);
      expect(component.selectedItems).toMatchSnapshot();
    });
  });

  describe('ChoicerComponent: Component or items are enabled.', () => {
    it('Neither Component nor single item are disabled, so single element should be selected', () => {
      setComponentState(false);
      setSingleItemState(false);
      component.onItemClicked(singleItem);
      expect(component.selectedItems).toMatchSnapshot();
    });

    it('Item clicked is selected, so it should be unselected when clicked', () => {
      setComponentState(false);
      setSingleItemState(false);
      setSelectedItem(singleItem);
      component.onItemClicked(singleItem);
      expect(component.selectedItems).toMatchSnapshot();
    });
  });

  describe('ChoicerComponent: ', () => {
    it('When write value is called, selectedItems are updated with the passed values', () => {
      setComponentState(false);
      setSingleItemState(false);
      component.writeValue(items);

      expect(component.selectedItems).toMatchSnapshot();
    });

    it('When registerOnChange is called passing a function, onChange function is set', () => {
      const mock = jest.fn();

      component.registerOnChange(mock);

      expect(component.onChange).toEqual(mock);
    });

    it('When registerOnTouched is called passing a function, onTouched function is set', () => {
      const mock = jest.fn();

      component.registerOnTouched(mock);

      expect(component.onTouched).toEqual(mock);
    });
  });
});


