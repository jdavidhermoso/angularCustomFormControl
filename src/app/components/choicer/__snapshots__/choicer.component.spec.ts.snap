// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChoicerComponent ChoicerComponent:  When 1`] = `
Array [
  Object {
    "description": "",
    "disabled": false,
    "id": 1,
    "image": "",
    "subtitle": "Item 1 subtitle",
    "title": "Item 1 title",
  },
]
`;

exports[`ChoicerComponent ChoicerComponent:  When write value is called, selectedItems are updated with the passed values 1`] = `
Array [
  Object {
    "description": "",
    "disabled": false,
    "id": 1,
    "image": "",
    "subtitle": "Item 1 subtitle",
    "title": "Item 1 title",
  },
]
`;

exports[`ChoicerComponent ChoicerComponent:  When write value is called, selectedItems are updated with the passed values 2`] = `Array []`;

exports[`ChoicerComponent ChoicerComponent: Component or items are disabled. Component is disabled, so none element should be selected 1`] = `Array []`;

exports[`ChoicerComponent ChoicerComponent: Component or items are disabled. Single item is disabled, so none element should be selected 1`] = `Array []`;

exports[`ChoicerComponent ChoicerComponent: Component or items are enabled. Item clicked is selected, so it should be unselected when clicked 1`] = `Array []`;

exports[`ChoicerComponent ChoicerComponent: Component or items are enabled. Neither Component nor single item are disabled, so single element should be selected 1`] = `
Array [
  Object {
    "description": "",
    "disabled": false,
    "id": 1,
    "image": "",
    "subtitle": "Item 1 subtitle",
    "title": "Item 1 title",
  },
]
`;
